Update: Fixes multiline no-warning-comments rule. (fixes #9884)

sstern6 opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[x ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
Updated the RegExp to look for the term provided in the body of the comment, not just the beginning. 

**Is there anything you'd like reviewers to focus on?**
Please review new test cases add/moved from valid to invalid.  The examples on the documentation were not throwing errors https://eslint.org/docs/2.0.0/rules/no-warning-comments as stated, ie:

[Previous to the PR was not throwing any errors](https://eslint.org/demo/#eyJ0ZXh0IjoiLypcbipVbmV4cGVjdGVkICdhbnkgb3RoZXIgdGVybScgY29tbWVudC5cbiAqIFRoZSBzYW1lIGdvZXMgZm9yIHRoaXMgVE9ETyBjb21tZW50XG4gKiBPciBhIGZpeG1lXG4gKiBhcyB3ZWxsIGFzIGFueSBvdGhlciB0ZXJtXG4qL1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo1LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnN0cnVjdG9yLXN1cGVyIjoyLCJuby1jYXNlLWRlY2xhcmF0aW9ucyI6Miwibm8tY2xhc3MtYXNzaWduIjoyLCJuby1jb21wYXJlLW5lZy16ZXJvIjoyLCJuby1jb25kLWFzc2lnbiI6Miwibm8tY29uc29sZSI6Miwibm8tY29uc3QtYXNzaWduIjoyLCJuby1jb25zdGFudC1jb25kaXRpb24iOjIsIm5vLWNvbnRyb2wtcmVnZXgiOjIsIm5vLWRlYnVnZ2VyIjoyLCJuby1kZWxldGUtdmFyIjoyLCJuby1kdXBlLWFyZ3MiOjIsIm5vLWR1cGUtY2xhc3MtbWVtYmVycyI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHktY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1lbXB0eS1wYXR0ZXJuIjoyLCJuby1lbXB0eSI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1peGVkLXNwYWNlcy1hbmQtdGFicyI6Miwibm8tbmV3LXN5bWJvbCI6Miwibm8tb2JqLWNhbGxzIjoyLCJuby1vY3RhbCI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zcGFyc2UtYXJyYXlzIjoyLCJuby10aGlzLWJlZm9yZS1zdXBlciI6Miwibm8tdW5kZWYiOjIsIm5vLXVuZXhwZWN0ZWQtbXVsdGlsaW5lIjoyLCJuby11bnJlYWNoYWJsZSI6Miwibm8tdW5zYWZlLWZpbmFsbHkiOjIsIm5vLXVuc2FmZS1uZWdhdGlvbiI6Miwibm8tdW51c2VkLWxhYmVscyI6Miwibm8tdW51c2VkLXZhcnMiOjIsIm5vLXVzZWxlc3MtZXNjYXBlIjoyLCJyZXF1aXJlLXlpZWxkIjoyLCJ1c2UtaXNuYW4iOjIsInZhbGlkLXR5cGVvZiI6Miwibm8td2FybmluZy1jb21tZW50cyI6Mn0sImVudiI6e319fQ==).
```
/*
*Unexpected 'any other term' comment.
 * The same goes for this TODO comment
 * Or a fixme
 * as well as any other term
*/
``` 

Only had 1 error not 2 errors:

```
// todo and fixme
```

Thank you

Fixes: https://github.com/eslint/eslint/issues/9884

platinumazure commented on lib/rules/no-warning-comments.js:
```diff
@@ -79,12 +81,12 @@ module.exports = {
             }
 
-            return new RegExp(prefix + escaped + suffix, "i");
+            return new RegExp(prefix + escaped + suffix + eitherOrWordBoundary + term + wordBoundary, "i");
```
I think it would be really helpful to have a comment here showing what the regex should look like for an example term.

Also: Based on the issues I've identified in the tests, we want to make sure that this regex is only augmented when `{ location: "anywhere" }` option is specified.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -24,15 +24,10 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "// any comment", options: [{ terms: ["fixme", "todo"] }] },
         "// any comment",
         { code: "// any comment", options: [{ location: "anywhere" }] },
-        { code: "// any comment with TODO, FIXME or XXX", options: [{ location: "start" }] },
```
This should still be valid: the rule option `{ location: "start" }` means only flag the comment if it starts with TODO/FIXME/XXX.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -24,15 +24,10 @@ ruleTester.run("no-warning-comments", rule, {
         "// any comment",
         { code: "// any comment", options: [{ location: "anywhere" }] },
-        { code: "// any comment with TODO, FIXME or XXX", options: [{ location: "start" }] },
-        "// any comment with TODO, FIXME or XXX",
```
This should be valid as well: `{ location: "start" }` appears to be the rule default. This should only be flagged if the `{ location: "anywhere" }` option is explicitly specified.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -24,15 +24,10 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "/* any block comment */", options: [{ terms: ["fixme", "todo"] }] },
         "/* any block comment */",
         { code: "/* any block comment */", options: [{ location: "anywhere" }] },
-        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }] },
```
This should be valid since the term is not at the start of the comment.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -24,15 +24,10 @@ ruleTester.run("no-warning-comments", rule, {
         "/* any block comment */",
         { code: "/* any block comment */", options: [{ location: "anywhere" }] },
-        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }] },
-        "/* any block comment with TODO, FIXME or XXX */",
```
This should be valid since the term is not at the start of the comment.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -24,15 +24,10 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "/* any block comment */", options: [{ location: "anywhere" }] },
-        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }] },
-        "/* any block comment with TODO, FIXME or XXX */",
-        "/* any block comment with (TODO, FIXME's or XXX!) */",
```
This should be valid since the term is not at the start of the comment.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -50,11 +45,15 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "/* any fixme or todo */", options: [{ terms: ["fixme", "todo"], location: "anywhere" }], errors: [{ message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'todo' comment." }] },
         { code: "/* any fixme or todo */", options: [{ location: "anywhere" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }] },
-        { code: "/* fixme and todo */", errors: [{ message: "Unexpected 'fixme' comment." }] },
+        { code: "/* fixme and todo */", errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }] },
```
This should only be invalid with 2 messages if `{ location: "anywhere" }` is specified as an option.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -50,11 +45,15 @@ ruleTester.run("no-warning-comments", rule, {
         { code: "/* eslint one-var: 2 */", options: [{ terms: ["eslint"] }], errors: [{ message: "Unexpected 'eslint' comment." }] },
-        { code: "/* eslint one-var: 2 */", options: [{ terms: ["one"], location: "anywhere" }], errors: [{ message: "Unexpected 'one' comment." }] }
+        { code: "/* eslint one-var: 2 */", options: [{ terms: ["one"], location: "anywhere" }], errors: [{ message: "Unexpected 'one' comment." }] },
+        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
```
This should only be invalid when `{ location: "anywhere" }` is specified.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -50,11 +45,15 @@ ruleTester.run("no-warning-comments", rule, {
-        { code: "/* eslint one-var: 2 */", options: [{ terms: ["one"], location: "anywhere" }], errors: [{ message: "Unexpected 'one' comment." }] }
+        { code: "/* eslint one-var: 2 */", options: [{ terms: ["one"], location: "anywhere" }], errors: [{ message: "Unexpected 'one' comment." }] },
+        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
+        { code: "/* any block comment with (TODO, FIXME's or XXX!) */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
```
This should only be invalid when `{ location: "anywhere" }` is specified.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -50,11 +45,15 @@ ruleTester.run("no-warning-comments", rule, {
+        { code: "/* eslint one-var: 2 */", options: [{ terms: ["one"], location: "anywhere" }], errors: [{ message: "Unexpected 'one' comment." }] },
+        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
+        { code: "/* any block comment with (TODO, FIXME's or XXX!) */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
+        { code: "/** \n *any block comment \n*with (TODO, FIXME's or XXX!) **/", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
```
This should only be invalid when `{ location: "anywhere" }` is specified.

platinumazure commented on tests/lib/rules/no-warning-comments.js:
```diff
@@ -50,11 +45,15 @@ ruleTester.run("no-warning-comments", rule, {
+        { code: "/* any block comment with TODO, FIXME or XXX */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
+        { code: "/* any block comment with (TODO, FIXME's or XXX!) */", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
+        { code: "/** \n *any block comment \n*with (TODO, FIXME's or XXX!) **/", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] },
+        { code: "// any comment with TODO, FIXME or XXX", options: [{ location: "start" }], errors: [{ message: "Unexpected 'todo' comment." }, { message: "Unexpected 'fixme' comment." }, { message: "Unexpected 'xxx' comment." }] }
```
This should only be invalid when `{ location: "anywhere" }` is specified.

platinumazure commented:
Oh, forgot to mention: This should be an "Update" since this is a bugfix that potentially increases the number of warnings the user might experience. Please change the commit message accordingly. Thanks!

sstern6 commented:
Thanks for the feedback @platinumazure addressed all comments and split the logic between location `start` and location `anywhere` for the appropriate regexp with some comments.  Let me know if there are any more comments to address

Thanks

kaicataldo commented:
Sorry for losing track of this. Marking this as accepted, since the corresponding issue is accepted. @sstern6 Do you mind updating the description to reference the issue?

@platinumazure have your comments been addressed?

platinumazure commented:
Apologies @sstern6 and @kaicataldo, yes, my concerns are addressed.

kaicataldo commented:
Thanks for contributing to ESLint!
